Public/Src/Cache/ContentStore/Distributed/Redis/RedisDatabaseAdapter.cs

```diff
@@ -286,16 +293,81 @@ public async Task<BoolResult> ExecuteBatchOperationAsync(Context context, IRedis
                         },
                         cancellationToken);
                     await batch.NotifyConsumersOfSuccess();
+
+                    RedisOperationSucceeded();
                     return BoolResult.Success;
                 }
                 catch (Exception ex)
                 {
                     batch.NotifyConsumersOfFailure(ex);
+                    HandleRedisExceptionAndResetMultiplexerIfNeeded(context, ex);
                     return new BoolResult(ex);
                 }
             }
         }
 
+        private void RedisOperationSucceeded()
+        {
+            // If the operation is successful, then reset the error count to 0.
+            Interlocked.Exchange(ref _connectionErrorCount, 0);
+        }
+
+        /// <summary>
+        /// This method "handles" redis connectivity errors and reset the connection multiplexer if there are no successful operations happening.
+        /// </summary>
+        private void HandleRedisExceptionAndResetMultiplexerIfNeeded(Context context, Exception exception)
+        {
+            if (IsRedisConnectionException(exception))
+            {
+                // Using double-checked locking approach to reset the connection multiplexer only once.
+                // Checking for greater then or equals because another thread can increment _connectionErrorCount.
+                if (Interlocked.Increment(ref _connectionErrorCount) >= _redisConnectionErrorLimit)

```

Comment:
>Interlocked.Increment(ref _connectionErrorCount) [](start = 20, length = 48)

I think we should not increment here since multiple threads can call this concurrently. Instead, we should only increment after resetting. Further, if another thread is currently resetting the connection, the current thread should wait, but not reset the connection.